File: desktop/electron/main/navigation.ts
import type {
  DesktopSettings,
  ExternalTarget,
} from "../shared/contract.ts";
import {
  normalizeControlPlaneUrl,
  normalizeLoopbackUrl,
} from "./validation.ts";
import { isValidQuickTunnelUrl } from "./tunnel.ts";

export function externalUrlFor(
  settings: DesktopSettings,
  target: ExternalTarget,
): URL {
  if (target === "preview") {
    return normalizeLoopbackUrl(settings.previewUrl);
  }
  const origin = normalizeControlPlaneUrl(settings.controlPlaneUrl);
  if (target === "dashboard") return new URL("/dashboard", origin);
  if (target === "login") return new URL("/login", origin);
  throw new Error("Cette destination externe n’est pas autorisée.");
}

export function quickTunnelPreviewUrl(raw: unknown): URL {
  if (!isValidQuickTunnelUrl(raw)) {
    throw new Error("Le lien public du tunnel est invalide ou inactif.");
  }
  return new URL(raw);
}

export function connectPreviewUrlFor(
  settings: DesktopSettings,
  rawTunnelUrl: unknown,
): URL {
  const tunnelUrl = quickTunnelPreviewUrl(rawTunnelUrl);
  const origin = normalizeControlPlaneUrl(settings.controlPlaneUrl);
  const destination = new URL("/connect-preview", origin);
  destination.hash = new URLSearchParams({
    url: tunnelUrl.toString(),
  }).toString();
  return destination;
}

File: desktop/src-tauri/src/navigation.rs
use crate::local_preview::normalize_loopback_url;
use crate::settings::{normalize_control_plane_url, read};
use tauri::AppHandle;
use tauri_plugin_opener::OpenerExt;

#[tauri::command]
pub fn open_external(app: AppHandle, target: String) -> Result<(), String> {
    let settings = read(&app)?;
    let url = match target.as_str() {
        "preview" => normalize_loopback_url(&settings.preview_url)?,
        "dashboard" => normalize_control_plane_url(&settings.control_plane_url)?
            .join("/dashboard")
            .map_err(|_| "Impossible de préparer l’adresse du dashboard.".to_string())?,
        "login" => normalize_control_plane_url(&settings.control_plane_url)?
            .join("/login")
            .map_err(|_| "Impossible de préparer l’adresse de connexion.".to_string())?,
        _ => return Err("Cette destination externe n’est pas autorisée.".into()),
    };

    app.opener()
        .open_url(url.as_str(), None::<&str>)
        .map_err(|error| format!("Impossible d’ouvrir le navigateur système : {error}"))
}
